alacritty_terminal/src/text_run.rs

```diff
@@ -0,0 +1,289 @@
+use crate::cursor::CursorKey;
+use crate::index::{Column, Line, Point};
+use crate::term::{
+    cell::{Flags, MAX_ZEROWIDTH_CHARS},
+    color::Rgb,
+    RenderableCell, RenderableCellContent,
+};
+
+#[derive(Debug)]
+struct RunStart {
+    line: Line,
+    column: Column,
+    fg: Rgb,
+    bg: Rgb,
+    bg_alpha: f32,
+    flags: Flags,
+}
+impl RunStart {
+    /// Compare cell and check if it belongs to the same run.
+    fn is_adjacent_to_cell(&self, rc: &RenderableCell) -> bool {
+        self.line == rc.line
+            && self.fg == rc.fg
+            && self.bg == rc.bg
+            && (self.bg_alpha - rc.bg_alpha).abs() < std::f32::EPSILON
+            && self.flags == rc.flags
+    }
+}
+
+#[derive(Debug)]
+pub enum TextRunContent {
+    Cursor(CursorKey),
+    CharRun(String, Vec<[char; MAX_ZEROWIDTH_CHARS]>),
+}
+
+/// Represents a set of renderable cells that all share the same rendering propreties.
+/// The assumption is that if two cells are in the same TextRun they can be sent off together to
+/// be shaped. This allows for ligatures to be rendered but not when something breaks up a ligature
+/// (e.g. selection hightlight) which is desired behavior.
+#[derive(Debug)]
+pub struct TextRun {
+    /// A run never spans multiple lines.
+    pub line: Line,
+    /// Span of columns the text run covers.
+    pub span: (Column, Column),
+    /// Cursor or sequence of characters.
+    pub content: TextRunContent,
+    /// Foreground color of text run content.
+    pub fg: Rgb,
+    /// Background color of text run content.
+    pub bg: Rgb,
+    /// Background color opacity of the text run
+    pub bg_alpha: f32,
+    /// Attributes of this text run.
+    pub flags: Flags,
+}
+
+impl TextRun {
+    fn from_iter_state(
```

Comment:
I would expect a new method to be pub and to have a signature akin to `(Line, (Column, Column), TextRunContent, Rgb, Rgb, f32, Flags) -> Self` or some subset. Some parameter set that clearly maps to the fields of text run perhaps with some reasonable defaults.

This method by contrast should not be used publicly and only makes sense in the context of the `TextRunIter`. I'm happy to change the name but I don't think it makes sense as a new method. Whatever the name is it should capture the semantics that this method constructs a text run inside the `TextRunIter` specifically from state the `TextRunIter` maintains, and shouldn't be public.

Reply:
I'm fine with not naming it `new`, I'm just not a huge fan of the name. It shouldn't really matter where the method is called from, but instead what it does. And I see it as a bit troublesome since there's no iterator anywhere to be seen in the method signature.

Reply:
I've moved it to TextRunIter and renamed it to `build_text_run`. The name should now match the behavior of the function and it's new location should clarify the scope of usage.